## import libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
df = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv', index_col=False)
path = r'C:\Users\amany\A4-Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'prepared data' , 'products_cleaned.csv'))
df_ords = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv')
df_deps = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\departments.csv')
orders_products_combined = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'orders_products_combined.pkl'))
df_prodcle = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\prepared data\orders_cleaned.csv')
df_ordcle = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\prepared data\products_cleaned.csv')
Ords_prods_merged_updated = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'Ords_prods_merged_updated.pkl'))
ords_prods_cust_merged = pd.read_pickle(os.path.join(path, '02 data' , 'prepared data' , 'ords_prods_cust_merged.pkl'))

In [4]:
ords_prods_cust_merged.shape

(32433514, 36)

### import data frame

In [2]:
path = r'C:\Users\amany\A4-Instacart Basket Analysis'
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))


### data Wrangling

In [3]:
df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [4]:
df.shape

(206209, 10)

In [5]:
# step 1:  renaming columns
df = df.rename(columns={ 
    'Surnam' : 'Last Name', 
    'STATE' : 'State', 
    'date_joined' : 'Date Joined', 
    'n_dependants' : 'Number of Dependents', 
    'fam_status' : 'Family Status', 
    'income' : 'Income'})

In [11]:
# Rename the column "User ID" to "user_id" for consistency and to follow naming conventions
df = df.rename(columns={'User ID' : 'user_id'})

In [7]:
df.head()

user_id First Name Last Name  Gender       State  Age Date Joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependents Family Status  Income  
0                     3       married  165665  
1                     0        single   59285  
2                     2       married   99568  
3                     0        single   42049  
4                     1       married   40374

In [8]:
df.describe()

user_id            Age  Number of Dependents         Income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

In [9]:
# step 2 finding the missing values
missing_values = df.isnull().sum()

In [10]:
missing_values

user_id                     0
First Name              11259
Last Name                   0
Gender                      0
State                       0
Age                         0
Date Joined                 0
Number of Dependents        0
Family Status               0
Income                      0
dtype: int64

First Name: 11,259 missing values – A significant number of entries in the First Name column are missing.

In [12]:
# handle missing values with a Placeholder like unknown
df['First Name'] = df['First Name'].fillna('Unknown')

In [14]:
# make sure the mssing values has been filled
missing_values = df.isnull().sum()

In [15]:
missing_values

user_id                 0
First Name              0
Last Name               0
Gender                  0
State                   0
Age                     0
Date Joined             0
Number of Dependents    0
Family Status           0
Income                  0
dtype: int64

In [16]:
# step 3 finding duplicate rows
duplicates = df.duplicated()

In [17]:
# count the total number of duplicates
num_duplicates = duplicates.sum()

In [18]:
num_duplicates

0

no duplicated rows 

In [19]:
# find the type of each column
df.dtypes


user_id                  int64
First Name              object
Last Name               object
Gender                  object
State                   object
Age                      int64
Date Joined             object
Number of Dependents     int64
Family Status           object
Income                   int64
dtype: object

### Q6: Combining dataframes

In [20]:
# importing dataframe
aggregated_ords_prods_merged_updated = pd.read_pickle(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\prepared data\aggregated_ords_prods_merged_updated.pkl')

In [21]:
# combining customer with aggregated_ords_prods_merged_updated
merged_df = pd.merge(aggregated_ords_prods_merged_updated, df, on='user_id', how='inner')

In [22]:
# checking merge
merged_df.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ...         frequency_flag First Name  Last Name  Gender  \
0          0  ...  Non_frequent_customer      Linda     Nguyen  Female   
1          1  ...  Non_frequent_customer      Linda     Nguyen  Female   
2          1  ...  Non_frequent_customer      Linda     Nguyen  Female   
3          1  ...  Non_frequent_customer      Linda     Nguyen  Female   
4          1  ...  Non_frequent_customer      Linda     Nguyen  Female   

     State Age Date Joined Number of Dependents Family Status Income  
0  Alabama  31   2/17/2019                    3       married  40423  
1  Alabama  31   2/17/2019                    3       married  40423  
2  Alabama  31   2/17/2019                    3       married  40423  
3  Alabama  31   2/17/2019                    3       married  40423  
4  Alabama  31   2/17/2019                    3       married  40423  

[5 rows x 35 columns]

In [23]:
merged_df.shape

(32433514, 35)

In [24]:
#exporting the new dataframe as a pickle
merged_df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_cust_merged.pkl'))